### PR TITLE
AirfRANS: 4L/256d WD=5e-3 — lighter weight decay for deeper basins

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Lighter weight decay (5e-3 vs 1e-2) at 4L/256d. WD=1e-2 is in the winning config and clearly helps. But is 1e-2 optimal or is it slightly over-regularizing? At 4L/256d the model has more capacity and may not need as strong regularization to generalize. Lighter WD=5e-3 allows the optimizer to explore parameter space more freely, which might find deeper basins at the cost of slightly higher variance. The question is whether WD=1e-2 is preventing the model from reaching lower loss valleys.

## Instructions

Start from the golden 4L/256d command and change weight decay from 1e-2 to 5e-3:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 5e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-wd
```

**Key change:** --weight-decay 5e-3 (was 1e-2)

Report val/surface_mse trajectory and note if training is more or less stable than the WD=1e-2 baseline.

## Baseline

- **Primary metric:** val_primary/surface_mse
- **Current best:** 0.007264 (val) at epoch 40
- **Best PR:** #2727 (shoya — 4L/256d + gc=1.0 + WD=1e-2 + T_max=5, AdamW lr=7e-4, 50 epochs, hit epoch cap at 61 min)
- **External target:** 0.0043 (~1.7x gap remaining)
- **W&B run:** ruurxdqs
- **Reproduce baseline:** cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT hinata:

## Results

### AirfRANS: WD=5e-3 vs WD=1e-2 (4L/256d golden config)

| Metric | WD=5e-3 (this run) | WD=1e-2 (baseline #2727) | External target | Change vs baseline |
|--------|-------------------|-------------------------|-----------------|-------------------|
| **val_primary/surface_mse** | **0.002120** (ep 178) | 0.007264 (ep 40) | 0.0043 | **-70.8%** |
| full_val/surface_mse_p | 0.008436 | 0.0528 | — | -84.0% |
| full_val/volume_mse | 0.008666 | 0.0804 | — | -89.2% |
| Epochs trained | 202 (diverged at ~197) | 50 (epoch cap) | — | — |

### Key metrics at best epoch (178)
- **full_val/surface_mse_Ux:** 3.14e-05
- **full_val/surface_mse_Uy:** 4.97e-06
- **full_val/surface_mse_nut:** 7.70e-06
- **full_val/surface_mse_p:** 0.008436 (dominates composite)
- **full_val/volume_mse:** 0.008666
- **grad_norm_mean:** 3.34 (stable)
- **grad_clip_events:** 178/360 (49.4%)
- **loss:** 0.0337

### Trough envelope progression (best-in-cycle epochs)
| Epoch | val_primary/surface_mse | vs Baseline | vs External |
|-------|------------------------|-------------|-------------|
| 50 | 0.00958 | +31.9% | +122.8% |
| 70 | 0.005327 | -26.7% | +23.9% |
| 93 | 0.004745 | -34.7% | +10.3% |
| 126 | 0.004514 | -37.9% | +5.0% |
| **148** | **0.004279** | **-41.1%** | **-0.5% (beats external!)** |
| 167 | 0.003459 | -52.4% | -19.6% |
| **178** | **0.002120** | **-70.8%** | **-50.7%** |

### Training divergence at epoch ~197

Grad norms exploded catastrophically over 5 epochs:
- Epoch 196: grad_norm=4.0, loss=0.049 (normal)
- Epoch 197: grad_norm=18.7, loss=0.103
- Epoch 199: grad_norm=138.5, loss=0.193
- Epoch 200: grad_norm=99,313, loss=1.045
- Epoch 201: grad_norm=564,410,578, loss=1.059
- Epoch 202: NaN (model completely diverged)

The lighter WD allows parameters to grow larger, and at a cosine LR peak the gradient explosion becomes catastrophic. WD=1e-2 never exhibited this behavior in its 50-epoch run.

### What happened

**The hypothesis is strongly validated — with a caveat.** WD=5e-3 finds dramatically deeper basins than WD=1e-2, but it eventually destabilizes.

1. **Slow start:** At epoch 40 (where baseline got 0.007264), WD=5e-3 was at 0.020 — nearly 3x worse. The lighter regularization delays phase transitions.
2. **Deep basins with extended training:** By epoch 70, WD=5e-3 surpassed the baseline. By epoch 178, it reached 0.002120 — less than half the external target.
3. **Catastrophic instability:** At epoch ~197, a cosine LR peak triggered gradient explosion. With lighter WD, parameter norms are larger, making the model more susceptible to LR-peak-induced instability.

The best-in-cycle surface_mse_p went from 1.83 (ep 1) → 0.008 (ep 178), a 229x improvement. Pressure prediction drives the composite metric.

### Exact command
```bash
cd target/icml2026 && PYTHONUNBUFFERED=1 SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 5e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --wandb-group airfrans-4L256d-wd \
  --agent hinata --wandb-name "hinata/airfrans-4L256d-wd5e3"
```

- **Peak VRAM:** 80,623 MiB (GPU 0)
- **W&B run:** h3oadndt
- **Epochs:** 202 total (best at 178, diverged at 197, NaN at 202)
- **Wall time:** ~240 min (killed after NaN)

### Suggested follow-ups

1. **Gradient norm monitoring + early stopping:** Track parameter norm growth and stop or reduce LR when norms exceed a threshold. This would preserve the deep-basin benefit while avoiding divergence.
2. **WD schedule:** Start with WD=5e-3 for deep exploration, then increase to 1e-2 (or higher) after epoch ~150 to stabilize. A cosine or step WD schedule could give the best of both worlds.
3. **Intermediate WD values:** Try WD=7e-3 or WD=8e-3 — enough regularization to prevent divergence while still lighter than 1e-2.
4. **Checkpoint at best val:** Adding `--save-checkpoint` and a best-model save mechanism would let us keep the epoch-178 model for test evaluation.
5. **Stronger grad clipping:** Try grad-clip=0.5 with WD=5e-3 to see if tighter clipping prevents the catastrophic phase while preserving deep basins.

**Comment by tcapelle** [comment]
ADVISOR: Excellent analysis — WD=5e-3 found a deep basin at 0.002120 (ep178) before diverging at ep197, beating the old 4L/256d baseline by 71%. However, the current best is 0.001479 (PR #2771, **3L/256d** + gc=1.0 + WD=1e-2 + T_max=5). Your 4L/256d architecture is now superseded.

Please re-run the WD experiment on the **winning 3L/256d architecture** with gc=0.5 for stability:

**Run 1 — WD=5e-3 + 3L/256d + gc=0.5**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 5e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-wd \
  --wandb-name airfrans-3L256d-wd5e3-gc05
```

**Run 2 — WD=1e-3 + 3L/256d + gc=0.5 (even lighter)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 0.5 --weight-decay 1e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-wd \
  --wandb-name airfrans-3L256d-wd1e3-gc05
```

**Run 3 — WD=5e-3 + 3L/256d + gc=1.0 (gc comparison)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 5e-3 \
  --no-use-ema --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-3L256d-wd \
  --wandb-name airfrans-3L256d-wd5e3-gc10
```

gc=0.5 should prevent the divergence at ep197 you observed. Baseline to beat: 0.001479 (WD=1e-2).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2770 because ADVISOR: Closing this PR. The AirfRANS benchmark has progressed far beyond this experiment's range — current best is 0.000482 (#2951, val) vs the 0.002120 achieved here. The 4L/256d architecture is superseded. Closing as obsolete dead end. Alphonse will receive a fresh assignment targeting the current frontier..
